File: chapter1/6.py
'''
Write a method to replace all spaces in a string with '%20'.
You may assume that the string has sufficient space at the end to
hold the additional characters, and that you are given the "true"
length of the string. (Note: If implementing in Java, please us a
character array so that you can perform this operation in place.
 
	EXAMPLE
 	
 	Input: "Mr John Smith    ", 13
 	Output: "Mr%20John%20Smith"

'''
def compress_string(s):
	last_character = None
	character_count = 1
	str2 = ""

	for c in s:
		if last_character == None:
			last_character = c
			str2 += c
		elif last_character==c:
			character_count += 1
		else:
			last_character = c
			str2 += str(character_count)
			character_count = 1
			str2 += c

	return str2

def main():
	print(compress_string("aaabeeffffg"))

if __name__ == "__main__":
	main()

File: chapter1/c++/6.cpp
/*
Implement a method to perform basic string compression using the counts of repeated characters. For example, the string aabcccccaaa would become a2b1c5a3. If the "compressed" string would not become smaller than the original string, your method should return the original string. You can assume the string has only uppercase and lowercase letters (a-z).
*/

#include <bits/stdc++.h>
using namespace std;

string compressString(string str) {
	char lastCharacter = '\0';
	int characterCount = 1;
	string str2 = "";

	for(int i=0; i<str.size(); i++) {
		if(lastCharacter=='\0') {
			lastCharacter = str[i];
			str2 += str[i];
		} else if(lastCharacter==str[i]) {
			characterCount++;
		} else {
			lastCharacter = str[i];
			str2 += to_string(characterCount);
			characterCount = 1;
			str2 += str[i];
		}
	}
	return str2;
}

int main() {
	cout << compressString("aaabeeffffg") << endl;
}
